### ISSUE TITLE
CPU usage for top n CPU

### ISSUE THREAD

**Issue by Yournished** [brndnmtthws/conky#695]
Nowadays processors have a lot of cores so it's not really practical trying to list every single one, 16 cores can take up a lot of space.

We can already get average of all cores with `cpu 0`. However, say that reports 12% usage - does that mean each core is running at 1/8th of its capacity, or that one core is being saturated by a single-threaded process? For cases like that it's useful to know how the usage is distributed across cores.

In this common case of one core being saturated, I think it would be sufficient to just show top 2 or 3 cores. The command can be `cputop n` to show the n'th currently most utilized core. If average is 12%, core 4 is 100%, core 7 is 10%, and all others are implicitly <10% (because the list is sorted) that lets the user tell in 3 graphs the same information they would get from 8 or 16.

**Comment by lasers** [comment]
I feel this could too tailored for everyday usages and wouldn't be largely utilized by the users. It might be better to use a combination of `{exec}` and custom scripts. With this in mind, you can write a script that will be 100% tailored to your needs. Thoughts on this feature request? @brndnmtthws

**Comment by brndnmtthws** [comment]
It's not a bad idea, and machines will probably have more CPUs in the future. I think we can just leave it open for now.

**Comment by Yournished** [comment]
>I feel this could too tailored for everyday usages and wouldn't be largely utilized by the users. 

I agree that it's a bit niche, but as brndnmtthws says, the number of cores on a machine will only increase going forward. To have usage graphs for individual cores at all is clearly not too tailored, because it was implemented in the first place. The current implementation is a fair solution for a small number of cores, but I suspect when it was implemented nobody thought about what happens when you have a dozen or more cores, because at that point the per-core graph feature becomes unusable. So I think the logical options here are to either kill the per core graph entirely, since it's too tailored, or to fix it so that it can deal with a situation where there are many cores but only a few of them are relevant at a time.

By the way, it occurs to me now that the standard way of dealing with this would be to have a graph with multiple lines, one for each core, and possibly each a different color. But because with conky it's common to use bar graphs rather than line, this might take a bit of work (not to mention the requirement for the legend if every core gets its own color). I think my solution of being able to show graphs of only the busiest cores is superior in terms of user experience, aesthetics and elegance. I imagine it wouldn't be that hard to implement, just sort the cpus by usage before returning the i'th cpu if a flag is set. A shame I'm not familiar enough with C to make a PR.

>It might be better to use a combination of `{exec}` and custom scripts. With this in mind, you can write a script that will be 100% tailored to your needs.

I could of course write my own script (I imagine I "could" also write my own conky alternative :-p). However the conky documentation has scary warnings about the supposed performance impact of using `exec`. I'm really only familiar with Python, would calling a python script every second to check CPU usage really work performance wise?

Also, I think writing a script for this isn't really practical. It would be easy to write a script that outputs a number for the usage of the n'th most busy core. But then even with 16 cores, printing 16 percentages does not take up much space, so the issue I bring up is not really relevant. It becomes a problem when you start printing the graphs. In which case it's quite non-trivial to write a script that would output a graph compatible with conky. Can this even be done? I've never tried. Does conky make graphs with arbitrary data?

**Comment by Yournished** [comment]
I tried to take a look at the source but I'm sad to say I got utterly lost. If this was Python, I would assume that somewhere there's a function like `get_cpu_usage(cpu_id)` that gets called when the `cpuN` parameter is given to `cpugraph`. If this was Python:

```
def get_cpu_usage(cpu_id):
    cpu_usage = get_all_cpu_usages()
    return cpu_usage[cpu_id]
```

could be changed to

```
def get_cpu_usage(cpu_id, busiest=False):
    cpu_usage = get_all_cpu_usages()
	
	if busiest:
		cpu_usage = sorted(cpu_usage)
	
	return cpu_usage[cpu_id]
```

Which function is called for `cpugraph` though, is it https://github.com/brndnmtthws/conky/blob/fcc315f28f384a71f23cd2f2140381f86dc841ee/src/cpu.cc#L125? That's a little bit more opaque than I'd hoped... Haven't seen goto in actual code in some time, I have to admit.
